README
This app goes through PAM with logind, forks into the user specified, like a login manager would do

usage

simplelogin --user USERNAME --exec FULLPATHTOEXEC

super simple sans display display desktop manager mockup mitigating mir

or sssdddmmmm for short.




Note
this file expects sddm-autologin in /etc/pam.d/
Use the ubuntu one, not the one from the repo. Ubuntu have a messed up system

renaming the lightdm one should be fine too

exec line can only be one string due to shoddy parsing, if you need to pass args, just go via a shell script or modify the code

debian/changelog
simplelogin (0.0.2+kubuntu1-0ubuntu1~ppa4) vivid; urgency=medium

  * bump

 -- Boudewijn Rempt <boud@kde.org>  Fri, 08 May 2015 11:54:13 +0000

simplelogin (0.0.2+kubuntu1-0ubuntu1~ppa3) vivid; urgency=medium

  * install kwinwrapper, not westonwrapper

 -- Boudewijn Rempt <boud@kde.org>  Fri, 08 May 2015 09:53:53 +0000

simplelogin (0.0.2+kubuntu1-0ubuntu1~ppa2) vivid; urgency=medium

  * add kwinwrapper

 -- Boudewijn Rempt <boud@kde.org>  Fri, 08 May 2015 08:30:23 +0000

simplelogin (0.0.1+kubuntu1-0ubuntu1~ppa1) vivid; urgency=medium

  * Initial release. 

 -- Boudewijn Rempt <boud@kde.org>  Wed, 06 May 2015 09:50:41 +0000


Backend.cpp
/*
 * Base backend class to be inherited further
 * Copyright (C) 2013 Martin B????za <mbriza@redhat.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */

#include "Backend.h"
#include "HelperApp.h"

#include "backend/PamBackend.h"
#include "UserSession.h"

#include <QtCore/QProcessEnvironment>

#include <pwd.h>

namespace SDDM {
    Backend *Backend::get(HelperApp* parent)
    {
        return new PamBackend(parent);
    }
    
    Backend::Backend(HelperApp* parent)
            : QObject(parent)
            , m_app(parent) {
    }

    void Backend::setAutologin(bool on) {
        m_autologin = on;
    }

    void Backend::setGreeter(bool on) {
        m_greeter = on;
    }

    bool Backend::openSession() {
        struct passwd *pw;
        pw = getpwnam(qPrintable(qobject_cast<HelperApp*>(parent())->user()));
        if (pw) {
            QProcessEnvironment env = m_app->session()->processEnvironment();
            env.insert("HOME", pw->pw_dir);
            env.insert("PWD", pw->pw_dir);
            env.insert("SHELL", pw->pw_shell);
            env.insert("USER", pw->pw_name);
            env.insert("LOGNAME", pw->pw_name);
            if (env.contains("DISPLAY") && !env.contains("XAUTHORITY"))
                env.insert("XAUTHORITY", QString("%1/.Xauthority").arg(pw->pw_dir));
            // TODO: I'm fairly sure this shouldn't be done for PAM sessions, investigate!
            m_app->session()->setProcessEnvironment(env);
        }
        return m_app->session()->start();
    }
}


HelperApp.h
/*
 * Main authentication application class
 * Copyright (C) 2013 Martin B????za <mbriza@redhat.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */

#ifndef Auth_H
#define Auth_H

#include <QtCore/QCoreApplication>
#include <QtCore/QProcessEnvironment>

namespace SDDM {
    class UserSession;
    class HelperApp : public QCoreApplication
    {
        Q_OBJECT
    public:
        HelperApp(int& argc, char** argv);
        virtual ~HelperApp();

        const QString &user() const;

    private Q_SLOTS:
        void sessionFinished(int status);

    private:
        qint64 m_id { -1 };
        UserSession *m_session { nullptr };
        QString m_user { };
    };
}

#endif // Auth_H


UserSession.cpp
/*
 * Session process wrapper
 * Copyright (C) 2014 Martin B????za <mbriza@redhat.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */

#include "UserSession.h"
#include "HelperApp.h"

#include <QDebug>

#include <sys/types.h>
#include <unistd.h>
#include <pwd.h>
#include <grp.h>
#include <unistd.h>
#include <fcntl.h>

namespace SDDM {
    UserSession::UserSession(HelperApp *parent)
            : QProcess(parent) {
    }

    UserSession::~UserSession() {

    }

    bool UserSession::start() {
        //NOTE FOR DAVE GO BACK AND FIX THIS IN SDDM, DON'T WAIT FOR STARTED IN HERE. IT'S WEIRD
        
        qDebug() << "starting "<< m_path;
        
        QProcess::start(m_path);

        return waitForStarted();
    }

    void UserSession::setPath(const QString& path) {
        m_path = path;
    }

    QString UserSession::path() const {
        return m_path;
    }

    void UserSession::setupChildProcess() {        
        QString username = qobject_cast<HelperApp*>(parent())->user();

        struct passwd *pw = getpwnam(username.toLocal8Bit());
        if (setgid(pw->pw_gid) != 0) {
            qCritical() << "setgid(" << pw->pw_gid << ") failed for user: " << username;
            exit(-1);
        }
        if (initgroups(pw->pw_name, pw->pw_gid) != 0) {
            qCritical() << "initgroups(" << pw->pw_name << ", " << pw->pw_gid << ") failed for user: " << username;
            exit(-1);
        }
        if (setuid(pw->pw_uid) != 0) {
            qCritical() << "setuid(" << pw->pw_uid << ") failed for user: " << username;
            exit(-1);
        }
        if (chdir(pw->pw_dir) != 0) {
            qCritical() << "chdir(" << pw->pw_dir << ") failed for user: " << username;
            qCritical() << "verify directory exist and has sufficient permissions";
            exit(-1);
        }

        //from here onwards we redirect stderr to a file, doesn't seem to work 
        return; 
        
                
        //we cannot use setStandardError file as this code is run in the child process
        //we want to redirect after we setuid so that .xsession-errors is owned by the user

        //swap the stderr pipe of this subprcess into a file .xsession-errors
        int fd = ::open(".xsession-errors-dave", O_WRONLY | O_CREAT | O_TRUNC, 0600);
        if (fd >= 0)
        {
            dup2 (fd, STDERR_FILENO);
            ::close(fd);
        } else {
            qWarning() << "Could not open stderr to .xsession-errors file";
        }

        //redirect any stdout to /dev/null
        fd = ::open("/dev/null", O_WRONLY);
        if (fd >= 0)
        {
            dup2 (fd, STDOUT_FILENO);
            ::close(fd);
        } else {
            qWarning() << "Could not redirect stdout";
        }
    }
}


backend/PamBackend.h
/*
 * PAM authentication backend
 * Copyright (C) 2013 Martin Bříza <mbriza@redhat.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */

#if !defined(PAMBACKEND_H) && defined(USE_PAM)
#define PAMBACKEND_H

// #include "Constants.h"
#include "../Backend.h"

#include <QtCore/QObject>

#include <security/pam_appl.h>

namespace SDDM {
    class PamHandle;
    class PamBackend;
    class PamData {
    public:
        PamData();

        bool insertPrompt(const struct pam_message *msg, bool predict = true);
        Auth::Info handleInfo(const struct pam_message *msg, bool predict);

        const Request& getRequest() const;
        void completeRequest(const Request& request);

        QByteArray getResponse(const struct pam_message *msg);

    private:
        AuthPrompt::Type detectPrompt(const struct pam_message *msg) const;

        const Prompt& findPrompt(const struct pam_message *msg) const;
        Prompt& findPrompt(const struct pam_message *msg);

        bool m_sent { false };
        Request m_currentRequest { };
    };

    class PamBackend : public Backend
    {
        Q_OBJECT
    public:
        explicit PamBackend(HelperApp *parent);
        virtual ~PamBackend();
        int converse(int n, const struct pam_message **msg, struct pam_response **resp);

    public slots:
        virtual bool start(const QString &user = QString());
        virtual bool authenticate();
        virtual bool openSession();

        virtual QString userName();

    private:
        PamData *m_data { nullptr };
        PamHandle *m_pam { nullptr };
    };
}

#endif // PAMBACKEND_H


backend/PamHandle.cpp
/*
 * PAM API Qt wrapper
 * Copyright (C) 2013 Martin Bříza <mbriza@redhat.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */
#include "PamHandle.h"
// #include "PamBackend.h"

#include <QtCore/QDebug>

namespace SDDM {
    bool PamHandle::putEnv(const QProcessEnvironment& env) {
        foreach (const QString& s, env.toStringList()) {
            m_result = pam_putenv(m_handle, qPrintable(s));
            if (m_result != PAM_SUCCESS) {
                qWarning() << "[PAM] putEnv:" << pam_strerror(m_handle, m_result);
                return false;
            }
        }
        return true;
    }

    QProcessEnvironment PamHandle::getEnv() {
        QProcessEnvironment env;
        // get pam environment
        char **envlist = pam_getenvlist(m_handle);
        if (envlist == NULL) {
            qWarning() << "[PAM] getEnv: Returned NULL";
            return env;
        }

        // copy it to the env map
        for (int i = 0; envlist[i] != nullptr; ++i) {
            QString s(envlist[i]);

            // find equal sign
            int index = s.indexOf('=');

            // add to the hash
            if (index != -1)
                env.insert(s.left(index), s.mid(index + 1));

            free(envlist[i]);
        }
        free(envlist);
        return env;
    }

    bool PamHandle::chAuthTok(int flags) {
        m_result = pam_chauthtok(m_handle, flags | m_silent);
        if (m_result != PAM_SUCCESS) {
            qWarning() << "[PAM] chAuthTok:" << pam_strerror(m_handle, m_result);
        }
        return m_result == PAM_SUCCESS;
    }

    bool PamHandle::acctMgmt(int flags) {
        m_result = pam_acct_mgmt(m_handle, flags | m_silent);
        if (m_result == PAM_NEW_AUTHTOK_REQD) {
            // TODO see if this should really return the value or just true regardless of the outcome
            return chAuthTok(PAM_CHANGE_EXPIRED_AUTHTOK);
        }
        else if (m_result != PAM_SUCCESS) {
            qWarning() << "[PAM] acctMgmt:" << pam_strerror(m_handle, m_result);
            return false;
        }
        return true;
    }

    bool PamHandle::authenticate(int flags) {
        qDebug() << "[PAM] Authenticating...";
        m_result = pam_authenticate(m_handle, flags | m_silent);
        if (m_result != PAM_SUCCESS) {
            qWarning() << "[PAM] authenticate:" << pam_strerror(m_handle, m_result);
        }
        qDebug() << "[PAM] returning.";
        return m_result == PAM_SUCCESS;
    }

    bool PamHandle::setCred(int flags) {
        m_result = pam_setcred(m_handle, flags | m_silent);
        if (m_result != PAM_SUCCESS) {
            qWarning() << "[PAM] setCred:" << pam_strerror(m_handle, m_result);
        }
        return m_result == PAM_SUCCESS;
    }

    bool PamHandle::openSession() {
        m_result = pam_open_session(m_handle, m_silent);
        if (m_result != PAM_SUCCESS) {
            qWarning() << "[PAM] openSession:" << pam_strerror(m_handle, m_result);
        }
        return m_result == PAM_SUCCESS;
    }

    bool PamHandle::closeSession() {
        m_result = pam_close_session(m_handle, m_silent);
        if (m_result != PAM_SUCCESS) {
            qWarning() << "[PAM] closeSession:" << pam_strerror(m_handle, m_result);
        }
        return m_result == PAM_SUCCESS;
    }

    bool PamHandle::setItem(int item_type, const void* item) {
        m_result = pam_set_item(m_handle, item_type, item);
        if (m_result != PAM_SUCCESS) {
            qWarning() << "[PAM] setItem:" << pam_strerror(m_handle, m_result);
        }
        return m_result == PAM_SUCCESS;
    }

    const void* PamHandle::getItem(int item_type) {
        const void *item;
        m_result = pam_get_item(m_handle, item_type, &item);
        if (m_result != PAM_SUCCESS) {
            qWarning() << "[PAM] getItem:" << pam_strerror(m_handle, m_result);
        }
        return item;
    }

    int PamHandle::converse(int n, const struct pam_message **msg, struct pam_response **resp, void *data) {
        qDebug() << "Dave assumed this wasn't needed and just removed a crucial bit of code. If you this it means that he was wrong (or you're using a stupid session which isn't autologin)";
//         PamBackend *c = static_cast<PamBackend *>(data);
//         return c->converse(n, msg, resp);
    }

    bool PamHandle::start(const QString &service, const QString &user) {
        if (user.isEmpty())
            m_result = pam_start(qPrintable(service), NULL, &m_conv, &m_handle);
        else
            m_result = pam_start(qPrintable(service), qPrintable(user), &m_conv, &m_handle);
        if (m_result != PAM_SUCCESS) {
            qWarning() << "[PAM] start" << pam_strerror(m_handle, m_result);
            return false;
        }
        else {
            qDebug() << "[PAM] Starting...";
        }
        return true;
    }

    bool PamHandle::end(int flags) {
        if (!m_handle)
            return false;
        m_result = pam_end(m_handle, m_result | flags);
        if (m_result != PAM_SUCCESS) {
            qWarning() << "[PAM] end:" << pam_strerror(m_handle, m_result);
            return false;
        }
        else {
            qDebug() << "[PAM] Ended.";
        }
        m_handle = NULL;
        return true;
    }

    QString PamHandle::errorString() {
        return pam_strerror(m_handle, m_result);
    }

    PamHandle::PamHandle(PamBackend *parent) {
        // create context
        m_conv = { &PamHandle::converse, parent };
    }

    PamHandle::~PamHandle() {
        // stop service
        end();
    }
}


backend/PamBackend.cpp
/*
 * PAM authentication backend
 * Copyright (C) 2013 Martin B????za <mbriza@redhat.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */

#include "PamBackend.h"
#include "PamHandle.h"
#include "HelperApp.h"
#include "UserSession.h"
#include "Auth.h"

#include <QtCore/QString>
#include <QtCore/QDebug>

#include <stdlib.h>

namespace SDDM {
    static Request loginRequest {
        {   { AuthPrompt::LOGIN_USER, "login:", false },
            { AuthPrompt::LOGIN_PASSWORD, "Password: ", true }
        }
    };

    static Request changePassRequest {
        {   { AuthPrompt::CHANGE_CURRENT, "(current) UNIX password: ", true },
            { AuthPrompt::CHANGE_NEW, "New password: ", true },
            { AuthPrompt::CHANGE_REPEAT, "Retype new password: ", true }
        }
    };

    static Request changePassNoOldRequest {
        {   { AuthPrompt::CHANGE_NEW, "New password: ", true },
            { AuthPrompt::CHANGE_REPEAT, "Retype new password: ", true }
        }
    };

    static Request invalidRequest { {} };

    static Prompt invalidPrompt {};

    PamData::PamData() { }

    AuthPrompt::Type PamData::detectPrompt(const struct pam_message* msg) const {
        if (msg->msg_style == PAM_PROMPT_ECHO_OFF) {
            QString message(msg->msg);
            if (message.indexOf(QRegExp("\\bpassword\\b", Qt::CaseInsensitive)) >= 0) {
                if (message.indexOf(QRegExp("\\b(re-?(enter|type)|again|confirm|repeat)\\b", Qt::CaseInsensitive)) >= 0) {
                    return AuthPrompt::CHANGE_REPEAT;
                }
                else if (message.indexOf(QRegExp("\\bnew\\b", Qt::CaseInsensitive)) >= 0) {
                    return AuthPrompt::CHANGE_NEW;
                }
                else if (message.indexOf(QRegExp("\\b(old|current)\\b", Qt::CaseInsensitive)) >= 0) {
                    return AuthPrompt::CHANGE_CURRENT;
                }
                else {
                    return AuthPrompt::LOGIN_PASSWORD;
                }
            }
        }
        else {
            return AuthPrompt::LOGIN_USER;
        }

        return AuthPrompt::UNKNOWN;
    }

    const Prompt& PamData::findPrompt(const struct pam_message* msg) const {
        AuthPrompt::Type type = detectPrompt(msg);

        for (const Prompt &p : m_currentRequest.prompts) {
            if (type == p.type && p.message == msg->msg)
                return p;
        }

        return invalidPrompt;
    }

    Prompt& PamData::findPrompt(const struct pam_message* msg) {
        AuthPrompt::Type type = detectPrompt(msg);

        for (Prompt &p : m_currentRequest.prompts) {
            if (type == AuthPrompt::UNKNOWN && msg->msg == p.message)
                return p;
            if (type == p.type)
                return p;
        }

        return invalidPrompt;
    }

    /*
    * Expects an empty prompt list if the previous request has been processed
    */
    bool PamData::insertPrompt(const struct pam_message* msg, bool predict) {
        Prompt &p = findPrompt(msg);

        // first, check if we already have stored this propmpt
        if (p.valid()) {
            // we have a response already - do nothing
            if (m_sent)
                return false;
            // we don't have a response yet - replace the message and prepare to send it
            p.message = msg->msg;
            return true;
        }
        // this prompt is not stored but we have some prompts
        else if (m_currentRequest.prompts.length() != 0) {
            // check if we have already sent this - if we did, get rid of the answers
            if (m_sent) {
                m_currentRequest.clear();
                m_sent = false;
            }
        }

        // we'll predict what will come next
        if (predict) {
            AuthPrompt::Type type = detectPrompt(msg);
            switch (type) {
                case AuthPrompt::LOGIN_USER:
                    m_currentRequest = Request(loginRequest);
                    return true;
                case AuthPrompt::CHANGE_CURRENT:
                    m_currentRequest = Request(changePassRequest);
                    return true;
                case AuthPrompt::CHANGE_NEW:
                    m_currentRequest = Request(changePassNoOldRequest);
                    return true;
                default:
                    break;
            }
        }

        // or just add whatever comes exactly as it comes
        m_currentRequest.prompts.append(Prompt(detectPrompt(msg), msg->msg, msg->msg_style == PAM_PROMPT_ECHO_OFF));

        return true;
    }

    Auth::Info PamData::handleInfo(const struct pam_message* msg, bool predict) {
        if (QString(msg->msg).indexOf(QRegExp("^Changing password for [^ ]+$"))) {
            if (predict)
                m_currentRequest = Request(changePassRequest);
            return Auth::INFO_PASS_CHANGE_REQUIRED;
        }
        return Auth::INFO_UNKNOWN;
    }

    /*
    * Destroys the prompt with that response
    */
    QByteArray PamData::getResponse(const struct pam_message* msg) {
        QByteArray response = findPrompt(msg).response;
        m_currentRequest.prompts.removeOne(findPrompt(msg));
        if (m_currentRequest.prompts.length() == 0)
            m_sent = false;
        return response;
    }

    const Request& PamData::getRequest() const {
        if (!m_sent)
            return m_currentRequest;
        else
            return invalidRequest;
    }

    void PamData::completeRequest(const Request& request) {
        if (request.prompts.length() != m_currentRequest.prompts.length()) {
            qWarning() << "[PAM] Different request/response list length, ignoring";
            return;
        }

        for (int i = 0; i < request.prompts.length(); i++) {
            if (request.prompts[i].type != m_currentRequest.prompts[i].type
                || request.prompts[i].message != m_currentRequest.prompts[i].message
                || request.prompts[i].hidden != m_currentRequest.prompts[i].hidden) {
                qWarning() << "[PAM] Order or type of the messages doesn't match, ignoring";
                return;
            }
        }

        m_currentRequest = request;
        m_sent = true;
    }




    PamBackend::PamBackend(HelperApp *parent)
            : Backend(parent)
            , m_data(new PamData())
            , m_pam(new PamHandle(this)) {
    }

    PamBackend::~PamBackend() {
        delete m_data;
        delete m_pam;
    }

    bool PamBackend::start(const QString &user) {
        bool result;

        QString service = "sddm";

        if (user == "sddm" && m_greeter)
            service = "sddm-greeter";
        else if (m_app->session()->path().isEmpty())
            service = "sddm-check";
        else if (m_autologin)
            service = "sddm-autologin";
        result = m_pam->start(service, user);

        if (!result)
            m_app->error(m_pam->errorString(), Auth::ERROR_INTERNAL);

        return result;
    }

    bool PamBackend::authenticate() {
        if (!m_pam->authenticate()) {
            m_app->error(m_pam->errorString(), Auth::ERROR_AUTHENTICATION);
            return false;
        }
        if (!m_pam->acctMgmt()) {
            m_app->error(m_pam->errorString(), Auth::ERROR_AUTHENTICATION);
            return false;
        }
        return true;
    }

    bool PamBackend::openSession() {
        if (!m_pam->setCred(PAM_ESTABLISH_CRED)) {
            m_app->error(m_pam->errorString(), Auth::ERROR_AUTHENTICATION);
            return false;
        }
        QProcessEnvironment sessionEnv = m_app->session()->processEnvironment();
        QString display = sessionEnv.value("DISPLAY");
        if (!display.isEmpty()) {
#ifdef PAM_XDISPLAY
            m_pam->setItem(PAM_XDISPLAY, qPrintable(display));
#endif
            m_pam->setItem(PAM_TTY, qPrintable(display));
        }
        if (!m_pam->putEnv(sessionEnv)) {
            m_app->error(m_pam->errorString(), Auth::ERROR_INTERNAL);
            return false;
        }
        if (!m_pam->openSession()) {
            m_app->error(m_pam->errorString(), Auth::ERROR_INTERNAL);
            return false;
        }
        sessionEnv.insert(m_pam->getEnv());
        m_app->session()->setProcessEnvironment(sessionEnv);
        return Backend::openSession();
    }

    QString PamBackend::userName() {
        return (const char*) m_pam->getItem(PAM_USER);
    }

    int PamBackend::converse(int n, const struct pam_message **msg, struct pam_response **resp) {
        qDebug() << "[PAM] Conversation with" << n << "messages";

        bool newRequest = false;

        if (n <= 0 || n > PAM_MAX_NUM_MSG)
            return PAM_CONV_ERR;

        for (int i = 0; i < n; i++) {
            switch(msg[i]->msg_style) {
                case PAM_PROMPT_ECHO_OFF:
                case PAM_PROMPT_ECHO_ON:
                    newRequest = m_data->insertPrompt(msg[i], n == 1);
                    break;
                case PAM_ERROR_MSG:
                    m_app->error(msg[i]->msg, Auth::ERROR_AUTHENTICATION);
                    break;
                case PAM_TEXT_INFO:
                    // if there's only the info message, let's predict the prompts too
                    m_app->info(msg[i]->msg, m_data->handleInfo(msg[i], n == 1));
                    break;
                default:
                    break;
            }
        }

        if (newRequest) {
            Request sent = m_data->getRequest();
            Request received;

            if (sent.valid()) {
                received = m_app->request(sent);

                if (!received.valid())
                    return PAM_CONV_ERR;

                m_data->completeRequest(received);
            }
        }

        *resp = (struct pam_response *) calloc(n, sizeof(struct pam_response));
        if (!*resp) {
            return PAM_BUF_ERR;
        }

        for (int i = 0; i < n; i++) {
            QByteArray response = m_data->getResponse(msg[i]);

            resp[i]->resp = (char *) malloc(response.length() + 1);
            // on error, get rid of everything
            if (!resp[i]->resp) {
                for (int j = 0; j < n; j++) {
                    free(resp[i]->resp);
                    resp[i]->resp = nullptr;
                }
                free(*resp);
                *resp = nullptr;
                return PAM_BUF_ERR;
            }

            memcpy(resp[i]->resp, response.constData(), response.length());
            resp[i]->resp[response.length()] = '\0';
        }

        return PAM_SUCCESS;
    }
}


HelperApp.cpp
/*
 * Main authentication application class
 * Copyright (C) 2013 Martin B????za <mbriza@redhat.com>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along
 * with this program; if not, write to the Free Software Foundation, Inc.,
 * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 *
 */

#include "HelperApp.h"
#include "UserSession.h"

#include "backend/PamHandle.h"

#include <QtCore/QTimer>
#include <QtCore/QFile>
#include <QtCore/QDebug>

#include <iostream>
#include <unistd.h>
#include <sys/socket.h>

namespace SDDM {
    HelperApp::HelperApp(int& argc, char** argv)
            : QCoreApplication(argc, argv)
            , m_session(new UserSession(this))
    {        
        QStringList args = QCoreApplication::arguments();
        int pos;
        

        if ((pos = args.indexOf("--exec")) >= 0) {
            if (pos >= args.length() - 1) {
                exit(-1);
            }
            m_session->setPath(args[pos + 1]);
        }

        if ((pos = args.indexOf("--user")) >= 0) {
            if (pos >= args.length() - 1) {
                exit(-1);
            }
            m_user = args[pos + 1];
        }
        
        if (m_session->path().isEmpty() || m_user.isEmpty()) 
            qFatal("pass some args please");
        
        PamHandle *pamHandle = new PamHandle; //TODO fix leak

        if (! pamHandle->start("sddm-autologin" /*PAM session*/, m_user)) //Martin check this exists
            qFatal("Could not start PAM");

        QProcessEnvironment env = m_session->processEnvironment();

//         env.insert("DISPLAY", name());
        env.insert("XDG_SEAT", "seat0");
//         env.insert("XDG_SEAT_PATH", daemonApp->displayManager()->seatPath(seat()->name()));
//         env.insert("XDG_SESSION_PATH", daemonApp->displayManager()->sessionPath(QString("Session%1").arg(daemonApp->newSessionId())));
        env.insert("XDG_VTNR", QString::number(1));
        env.insert("DESKTOP_SESSION", "KDE");
        env.insert("XDG_CURRENT_DESKTOP", "KDE");
        env.insert("XDG_SESSION_CLASS", "user");
        env.insert("XDG_SESSION_TYPE", "wayland");
//         env.insert("XDG_SESSION_DESKTOP", xdgSessionName);
        pamHandle->putEnv(env);

        pamHandle->setItem(PAM_TTY, "/dev/tty1");

        if (!pamHandle->authenticate())
            qFatal("Could not auth");

        if (!pamHandle->acctMgmt())
            qFatal("Could not do account management");

        if (!pamHandle->setCred(PAM_ESTABLISH_CRED))
            qFatal("Could not establish credentials");
        
        if (!pamHandle->openSession())
            qFatal("Could not open pam session");
        
        qDebug() << "startng process";
        
        connect(m_session, SIGNAL(finished(int)), SLOT(sessionFinished(int)));

        connect(m_session, &QProcess::readyReadStandardError, m_session, [this](){qDebug() << m_session->readAllStandardError();});
        connect(m_session, &QProcess::readyReadStandardOutput, m_session, [this](){qDebug() << m_session->readAllStandardOutput();});

        env.insert(pamHandle->getEnv());
        m_session->setProcessEnvironment(env);
        m_session->start();
        
    }

    void HelperApp::sessionFinished(int status) {
        exit(status);
    }
    
    const QString& HelperApp::user() const
    {
        return m_user;
    }


    HelperApp::~HelperApp() {

    }
}

int main(int argc, char** argv) {
    SDDM::HelperApp app(argc, argv);
    return app.exec();
}
